S3Proxy/cache.go

```diff
@@ -5,45 +5,104 @@ import (
 	"time"
 )
 
-var s3Buckets = map[string]*s3BucketCacheItem{}
+var s3Buckets = map[string]*S3BucketCacheItem{}
+var s3Objects = map[string]*S3ObjectCacheItem{}
 
-type s3BucketCacheItem struct {
-	Name      string
-	Location  string
+type s3CacheItem struct {
 	Timestamp time.Time
 	TTL       time.Duration
 }
 
-func (b s3BucketCacheItem) String() string {
-	tmp := &b
-	tmp.TTL = b.TTL - time.Since(b.Timestamp)
+func (c s3CacheItem) String() string {
+	tmp := &c
+	tmp.TTL = c.TimeLeft()
 	out, _ := json.Marshal(tmp)
 	return string(out)
 }
 
-func CacheBucketGet(name string) *s3BucketCacheItem {
+// TimeLeft returns the time duration before the CacheItem expires
+func (c s3CacheItem) TimeLeft() time.Duration {
+	return c.TTL - time.Since(c.Timestamp)
+}
+
+// S3BucketCacheItem is a cache entry representing bucket info
+type S3BucketCacheItem struct {
+	CacheItem s3CacheItem
+	Name      string
+	Location  string
+}
+
+func (b S3BucketCacheItem) String() string {
+	out, _ := json.Marshal(b)
+	return string(out)
+}
+
+// S3ObjectCacheItem represents a cache entry for meta data for an S3 object
+// and not the S3Object itself. The idea is to reduce S3 API calls for Object
+// information
+type S3ObjectCacheItem struct {
+	CacheItem s3CacheItem
+	Key       string
+	Bucket    string
+	FilePath  string
+}
+
+func (o S3ObjectCacheItem) String() string {
+	out, _ := json.Marshal(o)
+	return string(out)
+}
+
+// CacheBucketGet checks if the bucket information is in the local cache
+func CacheBucketGet(name string) *S3BucketCacheItem {
 	bucket, hit := s3Buckets[name]
 	if hit {
 		// We need to check that the cache entry hasn't expired
-		if time.Since(bucket.Timestamp) <= bucket.TTL {
+		if time.Since(bucket.CacheItem.Timestamp) <= bucket.CacheItem.TTL {
```

Comment:
You should be able to do `bucket.Timestamp` (I may be wrong, can't remember the exact semantics now)